src/dune/action_ast.ml

```diff
@@ -60,15 +60,29 @@ struct
           ; ( "with-accepted-exit-codes"
             , Dune_lang.Syntax.since Stanza.syntax (2, 0)
               >>> let+ codes = Predicate_lang.decode_one Dune_lang.Decoder.int
-                  and+ t = located t in
-                  match t with
-                  | _, ((Run _ | Bash _ | System _) as t) ->
+                  and+ loc, t = located t in
+                  let rec is_ok = function
```

Comment:
One thing that I forgot is that this check will need to be versioned: it should depend on the dune version in use.
If >= 2.2, it should be this check, otherwise just the nonrecursive check. You can find an example of how to check that [here](https://github.com/ocaml/dune/blob/77aa5d3dfdcd9a1a4bd1b0e26f8418b513760314/src/dune/dune_file.ml#L1615) (it might also be possible to add a `Syntax.since` call before recursing down).